File: Labs/Lab 5/ValidBST.py
class ValidBST:
    def isValidBST(self, root):
        return self.valid(root, float("-inf"), float("inf"))
    
    def valid(self, node, left, right):
        if node is None:
            return True
        if node.val <= left or node.val >= right:
            return False
        return (self.valid(node.left, left, node.val) and 
                self.valid(node.right, node.val, right))
    
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

vbst = ValidBST()
tree1 = TreeNode(4)
tree1.left=TreeNode(3)
tree1.left.left=TreeNode(1)
tree1.right=TreeNode(8)
tree1.right.left=TreeNode(5)
tree1.right.right=TreeNode(9)
tree2 = TreeNode(1)
tree2.left = TreeNode(2)
tree2.right = TreeNode(8)
tree2.right.left = TreeNode(3)
tree2.right.right = TreeNode(4)
tree2.right.left = TreeNode(5)
tree2.right.right = TreeNode(6)
print(vbst.isValidBST(tree1))
print(vbst.isValidBST(tree2))

File: Labs/Lab 5/ValidBST.java
public class ValidBST {
    public boolean isValidBST(TreeNode root) {
        return valid(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    public boolean valid(TreeNode node, long left, long right){
        if (node==null)
            return true;
        if (node.val<=left||node.val>=right)
            return false;
        return (valid(node.left,left, node.val)&&valid(node.right,node.val,right));
    }
    public static void main(String[] args){
        ValidBST vbst = new ValidBST();
        TreeNode t1 = new TreeNode(4);
        t1.left=new TreeNode(3);
        t1.left.left=new TreeNode(1);
        t1.right=new TreeNode(8);
        t1.right.left=new TreeNode(5);
        t1.right.right=new TreeNode(9);
        System.out.println(vbst.isValidBST(t1));

        TreeNode tree2 = new TreeNode(1);
        tree2.left = new TreeNode(2);
        tree2.right = new TreeNode(8);
        tree2.right.left = new TreeNode(3);
        tree2.right.right = new TreeNode(4);
        tree2.right.left = new TreeNode(5);
        tree2.right.right = new TreeNode(6);
        System.out.println(vbst.isValidBST(tree2));
    }

}
class TreeNode {

    int val;

    TreeNode left;

    TreeNode right;

    TreeNode() {}

    TreeNode(int val) { this.val = val; }

    TreeNode(int val, TreeNode left, TreeNode right) {

        this.val = val;

        this.left = left;

        this.right = right;

    }

 }